From 31d10f6cc394c0c2eb70da6d8cd577fcdf7e31e5 Mon Sep 17 00:00:00 2001
From: Portisch <hugo.portisch@yahoo.de>
Date: Fri, 21 Apr 2023 10:39:45 +0200
Subject: [PATCH 12/14] AMLCodec: on Dolby Vision playback stop wait until
 video is off The driver need first to clean up on video off.

---
 .../VideoPlayer/DVDCodecs/Video/AMLCodec.cpp  | 19 +++++++++++++++----
 1 file changed, 15 insertions(+), 4 deletions(-)

diff --git a/xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp b/xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp
index 6d7a346fb0..cdbfb693b2 100644
--- a/xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp
+++ b/xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp
@@ -2005,7 +2005,7 @@ bool CAMLCodec::OpenDecoder(CDVDStreamInfo &hints)
   if (dv_enable)
   {
     // enable Dolby Vision
-    CSysfsPath("/sys/module/amdolby_vision/parameters/dolby_vision_enable", 1);
+    CSysfsPath("/sys/module/amdolby_vision/parameters/dolby_vision_enable", 'Y');
 
     // force player led mode when enabled
     CSysfsPath dolby_vision_flags{"/sys/module/amdolby_vision/parameters/dolby_vision_flags"};
@@ -2030,7 +2030,7 @@ bool CAMLCodec::OpenDecoder(CDVDStreamInfo &hints)
   else if (device_support_dv)
   {
     // disable Dolby Vision
-    CSysfsPath("/sys/module/amdolby_vision/parameters/dolby_vision_enable", 0);
+    CSysfsPath("/sys/module/amdolby_vision/parameters/dolby_vision_enable", 'N');
   }
 
   // DEC_CONTROL_FLAG_DISABLE_FAST_POC
@@ -2290,8 +2290,19 @@ void CAMLCodec::CloseDecoder()
   // return tsync to default so external apps work
   CSysfsPath("/sys/class/tsync/enable", 1);
   // disable Dolby Vision driver
-  //CSysfsPath("/sys/module/amdolby_vision/parameters/dolby_vision_enable", 0);
-  // don't disable it as it would not switch back to non DV mode anymore
+  CSysfsPath dolby_vision_enable{"/sys/module/amdolby_vision/parameters/dolby_vision_enable"};
+  if (dolby_vision_enable.Exists() && StringUtils::EqualsNoCase(dolby_vision_enable.Get<std::string>(), "Y"))
+  {
+    CSysfsPath dv_video_on{"/sys/class/amdolby_vision/dv_video_on"};
+    if (dv_video_on.Exists())
+    {
+      std::chrono::time_point<std::chrono::system_clock> now(std::chrono::system_clock::now());
+      while(dv_video_on.Get<int>() == 1 && (std::chrono::system_clock::now() - now) < std::chrono::seconds(m_decoder_timeout))
+        usleep(10000); // wait 10ms
+    }
+    dolby_vision_enable.Set('N');
+  }
+
   CSysfsPath amdolby_vision_debug{"/sys/class/amdolby_vision/debug"};
   if (amdolby_vision_debug.Exists())
     amdolby_vision_debug.Set("enable_fel 0");
-- 
2.41.0

